compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===--------------- InterchangeTransposeGenericOps.cpp -------------------===//
+//
+// Interchange loops in generic ops to make the transpose happen on the outputs
+// instead of inputs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Transforms/FusionUtils.h"
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+struct TransposeGenericOpPattern : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!genericOp.hasTensorSemantics()) {
+      return rewriter.notifyMatchFailure(genericOp, "no tensor semantics");
+    }
+
+    // Pattern needs to trigger only on  elementwise ops.
+    if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) {
```

Comment:
We don't have to do it in the PR. The comments states that it's matching element-wise operation. Are we going to support other cases?

There are the checks about elementwise trait upstream. Maybe we can refactor some of those code and reuse them here.

https://github.com/llvm/llvm-project/blob/0252357b3e1e8f6d3bc51ac6d7ac075842b2c956/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L435-L448

Reply:
Thanks for the info. I will do it in a later PR.